src/overlay/OverlayManagerImpl.cpp

```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
 }
 
 void
-OverlayManagerImpl::storePeerList(std::vector<std::string> const& list,
-                                  bool setPreferred)
+OverlayManagerImpl::storePeerList(std::vector<PeerBareAddress> addresses,
+                                  bool setPreferred, bool startup)
 {
     auto typeUpgrade = setPreferred
                            ? PeerManager::TypeUpdate::SET_PREFERRED
                            : PeerManager::TypeUpdate::UPDATE_TO_OUTBOUND;
-
-    for (auto const& peerStr : list)
+    for (auto const& peer : addresses)
     {
-        try
+        if (setPreferred)
+        {
+            mConfigurationPreferredPeers.insert(peer);
+        }
+
+        if (startup)
         {
-            auto address = PeerBareAddress::resolve(peerStr, mApp);
-            getPeerManager().update(address, typeUpgrade,
+            getPeerManager().update(peer, typeUpgrade,
                                     PeerManager::BackOffUpdate::HARD_RESET);
         }
-        catch (std::runtime_error&)
+        else
         {
-            CLOG(ERROR, "Overlay") << "Unable to add peer '" << peerStr << "'";
+            // If address is present in the DB, `update` will ensure
+            // type is correctly updated. Otherwise, a new entry is created.
+            // Note that this won't downgrade promoted peers back to outbound.
+            getPeerManager().update(peer, typeUpgrade);
         }
     }
 }
 
 void
-OverlayManagerImpl::storeConfigPeers()
+OverlayManagerImpl::storeKnownAndPreferred(
+    std::vector<PeerBareAddress> known, std::vector<PeerBareAddress> preferred,
+    bool startup)
+{
+    storePeerList(std::move(known), false, startup);
+    storePeerList(std::move(preferred), true, startup);
+};
+
+void
+OverlayManagerImpl::storeConfigPeers(bool startup)
```

Comment:
This function has two branches that have nothing in common, so it will be better to just have two functions here (storeConfigPeersSync and storeConfigPeersAsync?).

Reply:
Sure, updated to `storeConfigPeers` and `triggerPeerResolution`.